lib/puppet/parser/functions/sprintf.rb

```diff
@@ -29,8 +29,22 @@
   :arity => -2,
   :doc => "Perform printf-style formatting of text.
 
-      The first parameter is format string describing how the rest of the parameters should be formatted.  See the documentation for the `Kernel::sprintf` function in Ruby for all the details."
+  The first parameter is format string describing how the rest of the parameters should be formatted.
+  See the documentation for the `Kernel::sprintf` function in Ruby for all the details."
 ) do |args|
-  fmt = args.shift
-  return sprintf(fmt, *args)
+  fmt = args[0]
+  args = args[1..-1]
+  begin
+    return sprintf(fmt, *args)
+  rescue KeyError => e
+    if args.size == 1 && args[0].is_a?(Hash)
```

Comment:
Wouldn't a better approach here be to use a different dispatcher for the one argument hash signature? That would require 4.x of course, but still. This bug has been sitting there for a while so perhaps worth it?

Reply:
cannot make this a 4.x function in a .z release